### DISCUSSION TITLE
Type hints for TransformManager

### DISCUSSION THREAD

**Discussion by AlexanderFabisch** [dfki-ric/pytransform3d#136] [Feature discussions]
@erikmannerfelt I checked how you use the TransformManager in [xdem](https://github.com/GlacioHack/xdem/blob/main/xdem/coreg.py#L1048). It seems like you are using integers as frame names. pytransform3d 1.9 will have type hints and my idea was to require frame names to be [strings](https://github.com/rock-learning/pytransform3d/blob/master/pytransform3d/transform_manager.pyi#L23). Would that be a problem? Would it be better for you if it could be str or int?

**Comment by erikmannerfelt** [comment]
Thanks @AlexanderFabisch for the consideration! For that particular case, I can just use string representations of the numbers, which is completely fine. I do however kind of like the integer indexing version, but it's nothing that can't be circumvented!

Alternatively, do you have a recommendation on how to make this more nicely?

**Reply by AlexanderFabisch** [reply]
Maybe you can explain a bit your use case. It seems like it is just a concatenation of multiple transforms? Maybe this is indeed something that is best solved with integer indices.

**Reply by AlexanderFabisch** [reply]
I come more from a robotics background and we often have similar cases (forward kinematics), but we typically have names for all the frames (links) in between.

**Reply by erikmannerfelt** [reply]
The use case there is for merging the transforms of a coregistration pipeline. Since there may be an arbitrary amount of steps in the pipeline, I opted for a single integer index. An example is:

1. A pipeline of elevation bias correction, followed by iterative closest point registration (two steps)
2. Fit the bias on the initial DEM (creates a vertical transform: A-B)
3. FIt the ICP transform on the bias-corrected DEM (creates a rigid 3D transform: B-C)
4. Export or apply the A-C transform.

Since there is no theoretical limit to the amount of steps involved, integer indices sounded good. But it could just as well be anything else unique (like the string representation of an integer). The only thing we're interested in there is the A-X transform, where X is the last step in the pipeline.

**Reply by AlexanderFabisch** [reply]
OK, I think it is best to allow `Union[str, int]` in this case to avoid unnecessary conversion between str and int. I will change that.

**Reply by AlexanderFabisch** [reply]
FYI, I will allow any hashable to be used as a frame name: https://github.com/rock-learning/pytransform3d/pull/137

**Reply by erikmannerfelt** [reply]
Sounds great! Thanks for your work!
